### PR TITLE
HADOOP-16699: Add verbose TRACE logging to ABFS

### PR DESCRIPTION
**Pull request by snvijaya**
The update aims to enable enhanced verbose logging at TRACE level which should in investigating thread hung scenarios.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 3109 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1085 | trunk passed |
| +1 | compile | 33 | trunk passed |
| +1 | checkstyle | 26 | trunk passed |
| +1 | mvnsite | 35 | trunk passed |
| +1 | shadedclient | 823 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 28 | trunk passed |
| 0 | spotbugs | 51 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 50 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 29 | the patch passed |
| +1 | compile | 25 | the patch passed |
| +1 | javac | 25 | the patch passed |
| -0 | checkstyle | 18 | hadoop-tools/hadoop-azure: The patch generated 1 new + 7 unchanged - 0 fixed = 8 total (was 7) |
| +1 | mvnsite | 28 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 816 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 23 | the patch passed |
| +1 | findbugs | 54 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 83 | hadoop-azure in the patch passed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 6401 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1712/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1712 |
| JIRA Issue | HADOOP-16699 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 4572337f1d45 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / df6b316 |
| Default Java | 1.8.0_222 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1712/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1712/1/testReport/ |
| Max. process+thread count | 454 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1712/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java [hunk 1]
```diff
@@ -62,6 +62,8 @@
 import org.apache.hadoop.security.ssl.DelegatingSSLSocketFactory;
 import org.apache.hadoop.security.ProviderUtils;
 import org.apache.hadoop.util.ReflectionUtils;
+import org.slf4j.Logger;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java]: separate import block above org.apache. thanks

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java [hunk 2]
```diff
@@ -533,7 +540,9 @@ public AccessTokenProvider getTokenProvider() throws TokenAccessProviderExceptio
         if (azureTokenProvider == null) {
           throw new IllegalArgumentException("Failed to initialize " + customTokenProviderClass);
         }
+        LOG.trace("Initializing " + customTokenProviderClass.getName());
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java]: use {} expansion

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java [hunk 3]
```diff
@@ -560,7 +569,9 @@ public AbfsAuthorizer getAbfsAuthorizer() throws IOException {
         @SuppressWarnings("unchecked")
         Class<AbfsAuthorizer> authClass = (Class<AbfsAuthorizer>) rawConfig.getClassByName(authClassName);
         authorizer = authClass.getConstructor(new Class[] {Configuration.class}).newInstance(rawConfig);
+        LOG.trace("Initializing " + authClassName);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java]: {}

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java [hunk 4]
```diff
@@ -560,7 +569,9 @@ public AbfsAuthorizer getAbfsAuthorizer() throws IOException {
...
         authorizer.init();
+        LOG.trace(authClassName + " init complete");
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AbfsConfiguration.java]: {} ... and anywhere else

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 5]
```diff
@@ -143,8 +143,12 @@ public AzureBlobFileSystemStore(URI uri, boolean isSecureScheme, Configuration c
     } catch (IllegalAccessException exception) {
       throw new FileSystemOperationUnhandledException(exception);
     }
+
+    LOG.trace("AbfsConfiguration init complete");
+
     this.userGroupInformation = UserGroupInformation.getCurrentUser();
     this.userName = userGroupInformation.getShortUserName();
+    LOG.trace("UGI init complete");
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: seems superfluous.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsRestOperation.java [hunk 6]
```diff
@@ -134,6 +134,8 @@ void execute() throws AzureBlobFileSystemException {
       throw new AbfsRestOperationException(result.getStatusCode(), result.getStorageErrorCode(),
           result.getStorageErrorMessage(), null, result);
     }
+
+    LOG.trace(operationType + " REST operation complete." );
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsRestOperation.java]: {}

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 5 above)
Comment (snvijaya) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Agree on the multiple logging, but a recent debugging session with an internal client revealed that there could be multiple scenarios which could look like a hung state in ABFSDriver initialization. The intial debugging culprits included SSLDelegationContext instance and UGI init hitting a system error and failing to return the control back. 
The actual cause of the issue turned out to be a hang in the initialization of the custom token provider used in the user scenario. 
As a precaution adding a log line whenever the control goes out of the ABFS driver code. As it is too verbose, adding as TRACE logs. Please let me know if you still feel that

**Comment by snvijaya** [comment]
@steveloughran - As mentioned in my comments above, please let me know your opinion if the log verbosity is undesired even though its at log level TRACE and is mostly into the initialize flow where its hard to locate any hung scenario. 

Test results with the latest changes:
On an account without Namespace enabled:
[INFO] Tests run: 52, Failures: 0, Errors: 0, Skipped: 0
[WARNING] Tests run: 395, Failures: 0, Errors: 0, Skipped: 211
[WARNING] Tests run: 192, Failures: 0, Errors: 0, Skipped: 126

On an account with Namespace enabled:
[INFO] Tests run: 52, Failures: 0, Errors: 0, Skipped: 0
[WARNING] Tests run: 395, Failures: 0, Errors: 0, Skipped: 25
[WARNING] Tests run: 192, Failures: 0, Errors: 0, Skipped: 126

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 36s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  18m  7s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 31s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 24s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 34s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  14m 37s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 21s |  trunk passed  |
| +0 :ok: |  spotbugs  |   0m 53s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 51s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 31s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 26s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 26s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 16s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 28s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  13m 35s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 22s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   0m 54s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 19s |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   |  55m 54s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1712/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1712 |
| JIRA Issue | HADOOP-16699 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux eb6e89595d77 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 8afabe4 |
| Default Java | 1.8.0_222 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1712/2/testReport/ |
| Max. process+thread count | 422 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1712/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 5 above)
Comment (steveloughran) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: well, if its in, print the username

**Comment by steveloughran** [comment]
If you are actually trying to debug the behaviour of plug-in components I would prefer that you embrace `org.apache.hadoop.util.DurationInfo` which can you log out in fellow debug at the start and end of all operations, and calculates actual durations -which is handy

example use

```java
try (DurationInfo d =
        new DurationInfo(LOG, false, "Waiting for task completion")) {
  future.join();
} catch (CompletionException e) {
  raiseInnerCause(e);
}
```

No matter the code flow, information will get logged. With your current patch if there is a failure nobody sees it.


You've been working here, if you're happy with it: fine. However, you can do more and for all your future patches I will be expecting it.

So: let me know what you want to do here?

1. Merge in as is
2. Explore DurationInfo

Your choice.

**Comment by snvijaya** [comment]
Thanks @steveloughran for the tip on DurationInfo. Was away on vacation and couldnt respond earlier. 
DurationInfo is very useful and we probably have more scenarios than the few initialization scenarios that this PR tries to add verbose logs for. Have created a separate task to track this :
https://issues.apache.org/jira/browse/HADOOP-16791
Will get to this after a couple of priority work items I am involved with currently.

Meanwhile, I would like this PR to get committed as additional updates would anyway be tracked in above JIRA. 
Could you please help commit this change.

**Comment by steveloughran** [comment]
well, provided you are aware that moving to DurationInfo will essentially require rolling back this logging, I am happy to merge it in.

Is this what everyone wants?

**Comment by snvijaya** [comment]
Hi @steveloughran , It could take few weeks before I can get back to this change. Meanwhile getting this change merged will help in any debugging and also also serve as pin points for the next change. 

Thanks a lot for your time.

**Comment by steveloughran** [comment]
Ok

+1, merged to trunk

**Pull request closed without merging**